File: F#/chap07/Chap0705.fs
namespace Chap0705

module C070501 =
    type UnvalidatedAddress = Undefined
    type CheckedAddress = Undefined
    type AddressValidationError = Undefined

    type AsyncResult<'success, 'failure> = Async<Result<'success, 'failure>>
    type CheckAddressExists = UnvalidatedAddress -> AsyncResult<CheckedAddress, AddressValidationError>

    type CheckProductCodeExists = Undefined
    type UnvalidatedOrder = Undefined
    type ValidatedOrder = Undefined
    type ValidationError = Undefined

    type ValidateOrder =
        CheckProductCodeExists // 依存関係
            -> CheckAddressExists // AsyncResult を返す依存関係
            -> UnvalidatedOrder
            -> AsyncResult<ValidatedOrder, ValidationError list>


module C070502 =
    open C070501

    type PricingError = PricingError of string
    type GetProductPrice = Undefined
    type PricedOrder = Undefined

    type PriceOrder =
        GetProductPrice // 依存関係
            -> ValidatedOrder // 入力
            -> Result<PricedOrder, PricingError>


module C070503 =
    open C070502

    type OrderAcknowledgement = Undefined
    type SendResult = Undefined
    type SendOrderAcknowledgment = OrderAcknowledgement -> Async<SendResult>

    type CreateOrderAcknowledgmentLetter = Undefined
    type OrderAcknowledgmentSent = Undefined

    type AcknowledgeOrder =
        CreateOrderAcknowledgmentLetter // 依存関係
            -> SendOrderAcknowledgment
            -> PricedOrder
            -> Async<OrderAcknowledgmentSent option>

File: Scala/chap07/src/main/scala/Chap0705.scala
import scala.concurrent.Future

object Chap070501:
  case class UnvalidatedAddress(Value: Any)

  case class CheckedAddress(Value: Any)

  case class AddressValidationError(Value: Any)

  type FutureEither[A, B] = Future[Either[A, B]]
  type CheckAddressExists = UnvalidatedAddress => FutureEither[AddressValidationError, CheckedAddress]

  case class CheckProductCodeExists(Value: Any)

  case class UnvalidatedOrder(Value: Any)

  case class ValidatedOrder(Value: Any)

  case class ValidationError(Value: Any)

  type ValidateOrder =
    CheckProductCodeExists // 依存
      => CheckAddressExists // 依存
      => UnvalidatedOrder // In
      => FutureEither[ValidationError, ValidatedOrder] // Out

object Chap070502:

  import Chap070501.ValidatedOrder

  case class PricingError(Value: String)

  case class GetProductPrice(Value: Any)

  case class PricedOrder(Value: Any)

  type PriceOrder =
    GetProductPrice // 依存
      => ValidatedOrder // In
      => Either[PricingError, PricedOrder] // Out

object Chap070503:

  import Chap070502.PricedOrder

  case class OrderAcknowledgement(Value: Any)

  case class SendResult(Value: Any)

  type SendOrderAcknowledgment = OrderAcknowledgement => Future[SendResult]

  case class CreateOrderAcknowledgmentLetter(Value: Any)

  case class OrderAcknowledgmentSent(Value: Any)

  type AcknowledgeOrder =
    CreateOrderAcknowledgmentLetter // 依存
      => SendOrderAcknowledgment // 依存
      => PricedOrder // In
      => Future[Option[OrderAcknowledgmentSent]] // Out
